### PR TITLE
Add configuration for bucket name and secured TLS connection

### LINKED ISSUE #116: Allow AWS S3 bucket for pipeline-log
## Expected Behaviour

Allow usage of AWS S3 bucket for pipeline-log

## Current Behaviour

Needs to accept:
- TLS flag
- custom "bucket name"
- possibly "region" too

## Possible Solution

Enable two more env-vars to configure bucket name and whether to set TLS.

Code:

https://github.com/openfaas/openfaas-cloud/blob/master/pipeline-log/handler.go

### PR DESCRIPTION
**Pull request by martindekov**
Add two environmental variables bucket_name and
secure_connection which enable customisation by default
connection is false and bucket name is pipeline

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

## Description

Added two environmental variables `s3_bucket` default set to `pipeline` and `s3_tls` default set to false in `gateway_config.yml` to enable customisation which are handled in pipeline-log function.

Closes #116 

## How Has This Been Tested?

With @bartsmykla help I was able to have S3 bucket named `openfaas-martin` and added a little tweak to the code that I removed from the PR so the `pipeline-logs` logs show the variables.

Working case:
set `pipeline-log` variables in stack.yml like so 
```
  pipeline-log:
    lang: go
    handler: ./pipeline-log
    image: martindekov/pipeline-log:0.3.1
    environment:
      write_debug: true
      read_debug: true
      combine_output: false
      secure_connection: false #new variable
      bucket_name: openfaas-martin #new variable
    environment_file:
      - gateway_config.yml
    secrets:
      - s3-access-key
      - s3-secret-key
      - payload-secret
```
Pushing function gocloudrepo triggered `pipeline-log` with following logs(the little tweaks mentioned above)
```
...
pipeline-log.1.zbfkocxb54va@FaasSwarm    | 2018/09/04 14:21:58 stderr: 2018/09/04 14:21:58 Bucket name is set to openfaas-martin
pipeline-log.1.zbfkocxb54va@FaasSwarm    | 2018/09/04 14:21:58 endpoint is s3.amazonaws.com
...
```
and the log was loaded into the `openfaas-martin` bucket in S3 as a folder with the name of my function-`gocloudfunc`

Second case comment out bucketname like
```
  pipeline-log:
...
      #bucket_name: openfaas-martin
...
```
The logs show 
```
...
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:45 stderr: 2018/09/04 14:33:44 Bucket name not found, set to default: pipeline
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:44 Bucket name is set to pipeline
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:44 endpoint is s3.amazonaws.com
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:45 error writing: pipeline/martindekov/gocloudfunc/8e0ca7c3007b45c7ab426679c28a2c5189234fee/gocloudfunc/build.log, error: Access Denied
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:45 Success=false, Error=exit status 1
pipeline-log.1.1k9hzriyp9db@FaasSwarm    | 2018/09/04 14:33:45 Out=
...
```
Since the bucket name was missing it fell back to the default one `pipeline` for minio cases and it failed to deploy my logs

Third case commented out  `secure_connection` so false by default and `bucket_name: openfaas-martin` 

```
  pipeline-log:
...
      #secure_connection: false #new variable
      bucket_name: openfaas-martin
...
```
so again the log was updated to the S3 bucket and showed in `system-pipeline`
```
...
pipeline-log.1.pg25xwj1npib@FaasSwarm    | 2018/09/04 15:05:27 stderr: 2018/09/04 15:05:26 Bucket name is set to openfaas-martin
pipeline-log.1.pg25xwj1npib@FaasSwarm    | 2018/09/04 15:05:26 endpoint is s3.amazonaws.com
pipeline-log.1.pg25xwj1npib@FaasSwarm    | 2018/09/04 15:05:27 Duration: 0.642165 seconds
pipeline-log.1.pg25xwj1npib@FaasSwarm    | Wrote 14215 bytes to openfaas-martin/martindekov/gocloudfunc/8e0ca7c3007b45c7ab426679c28a2c5189234fee/gocloudfunc/build.log
...
```

When I specify bucket that does not exists the logs show:
```
...
pipeline-log.1.nn64lfa44bw7@FaasSwarm    | 2018/09/05 09:14:56 stderr: 2018/09/05 09:14:56 error writing: watisgoinon/martindekov/gocloudfunc/b8f19b37d4a987c5c7864bb56bb653dd7872f259/gocloudfunc/build.log, error: The specified bucket does not exist

...
```

The commits are the same since I redeliver payload and the bucket is private so I cant paste it here.

## Checklist:

I have:

- [ ] updated the documentation if required
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

### CODE REVIEW

File: pipeline-log/handler.go [hunk 1]
```diff
@@ -92,14 +96,23 @@ func connectToMinio(region string) (*minio.Client, error) {
 
 	endpoint := os.Getenv("s3_url")
 
+	secured := secureConEnabled()
```
Comment (alexellis) [inline pipeline-log/handler.go]: secureConEnabled -> tlsEnabled.

File: pipeline-log/handler.go [hunk 2]
```diff
@@ -92,14 +96,23 @@ func connectToMinio(region string) (*minio.Client, error) {
...
 }
 
 // getPath produces a string such as pipeline/alexellis/super-pancake-fn/commit-id/fn1/
 func getPath(bucket string, p *sdk.PipelineLog) string {
 	fileName := "build.log"
 	return fmt.Sprintf("%s/%s/%s/%s/%s", bucket, p.RepoPath, p.CommitSHA, p.Function, fileName)
 }
+
+func secureConEnabled() bool {
+	if connection := os.Getenv("secure_connection"); connection == "true" || connection == "1" {
```
Comment (alexellis) [inline pipeline-log/handler.go]: Use the same naming pattern please. I.e. `s3_tls_enabled` or `s3_tls`

File: stack.yml [hunk 3]
```diff
@@ -139,6 +139,8 @@ functions:
       write_debug: true
       read_debug: true
       combine_output: false
+      secure_connection: false
```
Comment (alexellis) [inline stack.yml]: These should be in the same file as the ones containing `s3_url`. You also need to update the README.md to give AWS-specific instructions.

File: docs/README.md [hunk 4]
```diff
@@ -298,15 +298,24 @@ Logs from the container builder are stored in S3. This can be Minio which is S3-
...
-> Note: AWS S3 needs an additional two flags to be implemented to make sure TLS and a custom bucket name are used.
-
 * Generate secrets for Minio
 
 ```
 SECRET_KEY=$(head -c 12 /dev/urandom | shasum| cut -d' ' -f1)
 ACCESS_KEY=$(head -c 12 /dev/urandom | shasum| cut -d' ' -f1)
 ```
 
+For S3 we need to create access and security keys in AWS the following way:
```
Comment (alexellis) [inline docs/README.md]: If you'd prefer to use an S3 Bucket hosted on AWS.

File: docs/README.md [hunk 5]
```diff
@@ -298,15 +298,24 @@ Logs from the container builder are stored in S3. This can be Minio which is S3-
...
+
+* Generate access key by using the security credentials page. Expand the Access Keys section, and then Create New Root Key.
+
+* Generate secret key by opening the IAM console. Choose Users in the Details pane, pick the IAM user which will use the keys, and then Create Access Key on the Security Credentials tab.
```
Comment (alexellis) [inline docs/README.md]: > Generate <the>.

File: docs/README.md [hunk 6]
```diff
@@ -354,10 +363,21 @@ docker service create --constraint="node.role==manager" \
...
 ```
 
-Enter the value of the DNS above into `s3_url` in `gateway_config.yml` adding the port at the end:`minio:9000`
+Minio: 
+
+* Enter the value of the DNS above into `s3_url` in `gateway_config.yml` adding the port at the end:`minio:9000`
 
 > Note: For debugging and testing. You can expose the port of Minio with `docker service update minio --publish-add 9000:9000`, but this is not recommended on the public internet.
 
+S3:
```
Comment (alexellis) [inline docs/README.md]: S3 is a standard not a type of S3 - be explicit - Minio or AWS S3.

File: docs/README.md [hunk 7]
```diff
@@ -354,10 +363,21 @@ docker service create --constraint="node.role==manager" \
...
+
+* Enter the value of the DNS `s3.amazonaws.com` into `s3_url` in `gateway_config.yml`
+
+* In the same file set `s3_tls: true` and `s3_bucket` to the bucket you created in S3 like `s3_bucket: example-bucket`
+
+* Check where your bucket region is located and set `s3_region` accordingly
```
Comment (alexellis) [inline docs/README.md]: Simply:

> Update the `s3_region` value such as : (give the example for the one you used)

File: pipeline-log/handler.go [hunk 8]
```diff
@@ -92,14 +96,23 @@ func connectToMinio(region string) (*minio.Client, error) {
 
 	endpoint := os.Getenv("s3_url")
 
+	secured := tlsEnabled()
```
Comment (alexellis) [inline pipeline-log/handler.go]: `secured` -> `tlsEnabled`

File: pipeline-log/handler_test.go [hunk 9]
```diff
@@ -18,3 +19,34 @@ func Test_getPath(t *testing.T) {
 		t.Fail()
 	}
 }
+
+func Test_tlsEnabled(t *testing.T) {
```
Comment (alexellis) [inline pipeline-log/handler_test.go]: LGTM, but what about the other flags?

Comment (martindekov) [reply pipeline-log/handler_test.go]: Ah I will add tests for all existing s3 flags.

Comment (martindekov) [reply pipeline-log/handler_test.go]: Moved bucketName to a function `bucketName() string` and made 1 additional test. I am not sure for `region` though. Have a default value? Exit when region does not exist ?

**Comment by martindekov** [comment]
Also tested the latest changes on my OF-cloud instance just in case

**Pull request merged**